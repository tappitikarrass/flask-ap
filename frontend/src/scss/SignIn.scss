@use "main";

#signin {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

// desktop
@media screen and (min-width: 810px) {
  #page {
    width: 30vw;
    margin-left: 35vw;
  }
}

// tablet
@media screen and (min-width: 500px) and (max-width: 810px) {
  #page {
    width: 50vw;
    margin-left: 25vw;
  }
}
