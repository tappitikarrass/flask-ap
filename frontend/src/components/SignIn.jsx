import React, { useRef, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { useCookies } from 'react-cookie';
import { useForm } from 'react-hook-form';
import Notify from './Notify';
import '../scss/SignIn.scss';

function SignIn() {
  const navigate = useNavigate();
  const [cookies, setCookie, removeCookie] = useCookies(['token', 'username']);
  const loginForm = useRef();
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm(
    {
      defaultValues: {
        username: 'asket',
        password: '12345678',
      },
    },
  );

  useEffect(() => {
    if (cookies.token !== 'undefined') {
      navigate('/profile');
    }
  });

  async function onSubmit() {
    const userData = new FormData(loginForm.current);
    const encodedCreds = window.btoa(
      `${userData.get('username')}:${userData.get('password')}`,
    ).toString('base64');
    const response = await fetch(
      'http://localhost/backend/login',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Basic ${encodedCreds}`,
        },
      },
    );
    if (response.status === 200) {
      if (cookies.token != null) {
        removeCookie('token');
        removeCookie('username');
      }
      const tokenJson = await response.json();
      setCookie(
        'token',
        tokenJson.token,
        {
          path: '/',
          sameSite: 'Lax',
        },
      );
      setCookie(
        'username',
        userData.get('username'),
        {
          path: '/',
          sameSite: 'Lax',
        },
      );
      navigate('/profile');
    } else {
      document.getElementById('login-error').style.opacity = 1;
      setTimeout(() => {
        document.getElementById('login-error').style.opacity = 0;
      }, 2000);
    }
  }

  return (
    <div id="signin" className="content">
      <Notify
        id="login-error"
        type="error"
        title="Login error"
        msg="Wrong username or password"
      />
      <h2>Sign In</h2>
      <form
        id="signin-form"
        ref={loginForm}
        method="post"
        acceptCharset="utf-8"
        onSubmit={handleSubmit(onSubmit)}
      >
        <input
          type="text"
          {...register('username', { required: 'Username is required.' })}
          placeholder="Username"
          className="field"
        />
        <p>{errors.username?.message}</p>
        <input
          type="password"
          {...register('password', { required: 'Password is required.' })}
          placeholder="Password"
          className="field"
        />
        <p>{errors.password?.message}</p>
        <button id="login-bt" type="submit" className="bt-blue-end">Login</button>
      </form>
    </div>
  );
}

export default SignIn;
