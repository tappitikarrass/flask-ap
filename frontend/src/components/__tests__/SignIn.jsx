import React from "react";
import { HashRouter } from "react-router-dom";
import { create, act } from "react-test-renderer";
import SignIn from '../SignIn';

describe('SignIn', () => {
  it('',() => {
    let component;
    act(() => {
      component = create(
        <HashRouter>
          <SignIn />
        </HashRouter>
      );
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});
