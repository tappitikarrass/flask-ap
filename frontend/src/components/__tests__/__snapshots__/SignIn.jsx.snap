// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn  1`] = `
<div
  className="content"
  id="signin"
>
  <div
    className="notify-error"
    id="login-error"
  >
    <div
      className="notify-text"
    >
      <h3
        className="notify-title"
      >
        Login error
      </h3>
      <p
        className="notify-description"
      >
        Wrong username or password
      </p>
    </div>
  </div>
  <h2>
    Sign In
  </h2>
  <form
    acceptCharset="utf-8"
    id="signin-form"
    method="post"
    onSubmit={[Function]}
  >
    <input
      className="field"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Username"
      type="text"
    />
    <p />
    <input
      className="field"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Password"
      type="password"
    />
    <p />
    <button
      className="bt-blue-end"
      id="login-bt"
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
